pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -394,75 +406,170 @@ func (s *Scheduler) updateCachedPodData(p *corev1.Pod) {
-				continue
-			} else if len(nodeClaimTemplate.InstanceTypeOptions) != len(instanceTypes) {
+		if remaining, ok := s.remainingResources[s.nodeClaimTemplates[i].NodePoolName]; ok {
+			it = filterByRemainingResources(it, remaining)
+			if len(it) == 0 {
+				errs[i] = serrors.Wrap(fmt.Errorf("all available instance types exceed limits for nodepool"), "NodePool", klog.KRef("", s.nodeClaimTemplates[i].NodePoolName))
+				return
+			} else if len(s.nodeClaimTemplates[i].InstanceTypeOptions) != len(it) {
 				log.FromContext(ctx).V(1).WithValues(
-					"NodePool", klog.KRef("", nodeClaimTemplate.NodePoolName),
+					"NodePool", klog.KRef("", s.nodeClaimTemplates[i].NodePoolName),
 				).Info(fmt.Sprintf(
 					"%d out of %d instance types were excluded because they would breach limits",
-					len(nodeClaimTemplate.InstanceTypeOptions)-len(instanceTypes),
-					len(nodeClaimTemplate.InstanceTypeOptions),
+					len(s.nodeClaimTemplates[i].InstanceTypeOptions)-len(it),
+					len(s.nodeClaimTemplates[i].InstanceTypeOptions),
 				))
 			}
 		}
+		nodeClaim := NewNodeClaim(s.nodeClaimTemplates[i], s.topology, s.daemonOverhead[s.nodeClaimTemplates[i]], s.daemonHostPortUsage[s.nodeClaimTemplates[i]], it, s.reservationManager, s.reservedOfferingMode)
+		r, it, ofs, err := nodeClaim.CanAdd(ctx, pod, s.cachedPodData[pod.UID])
+		if err != nil {
+			errs[i] = err
 
-		nodeClaim := NewNodeClaim(nodeClaimTemplate, s.topology, s.daemonOverhead[nodeClaimTemplate], s.daemonHostPortUsage[nodeClaimTemplate], instanceTypes, s.reservationManager, s.reservedOfferingMode)
-		if err := nodeClaim.Add(ctx, pod, s.cachedPodData[pod.UID]); err != nil {
-			nodeClaim.Destroy()
+			// If the pod is compatible with a NodePool with reserved offerings available, we shouldn't fall back to a NodePool
+			// with a lower weight. We could consider allowing "fallback" to NodePools with equal weight if they also have
+			// reserved capacity in the future if scheduling latency becomes an issue.
 			if IsReservedOfferingError(err) {
-				errs = multierr.Append(errs, fmt.Errorf(
-					"compatible with nodepool %q but failed to add pod while adhering to reservation fallback policy, %w",
-					nodeClaimTemplate.NodePoolName,
-					err,
-				))
-				// If the pod is compatible with a NodePool with reserved offerings available, we shouldn't fall back to a NodePool
-				// with a lower weight. We could consider allowing "fallback" to NodePools with equal weight if they also have
-				// reserved capacity in the future if scheduling latency becomes an issue.
-				break
+				mu.Lock()
+				defer mu.Unlock()
+
+				// A reserved offering error means that any subsequent successful after this NodeClaimTemplate isn't valid
+				if i >= idx {
+					return
+				}
+				newNodeClaim = nil
+				requirements = nil
+				instanceTypes = nil
+				offeringsToReserve = nil
+				idx = i
+				cancel() // Cancel the context to end the workqueue parallelization
```

Comment:
I believe this is safe to cancel because the workqueue implementation guarantees that all calls for previous values of `i` (and potentially some future values) have already been invoked and won't be interrupted with the context. This isn't very clear to me without checking the workqueue implementation so if we stick with this implementation rather than implementing our own, I think it's worth a comment.